pkg/backup/item_collector.go

```diff
@@ -532,28 +705,72 @@ func (r *itemCollector) listItemsForLabel(unstructuredItems []unstructured.Unstr
 }
 
 // backupNamespaces process namespace resource according to namespace filters.
-func (r *itemCollector) backupNamespaces(unstructuredList *unstructured.UnstructuredList,
-	ie *collections.IncludesExcludes, gr schema.GroupResource, preferredGVR schema.GroupVersionResource,
-	log logrus.FieldLogger) []*kubernetesResource {
-	var items []*kubernetesResource
-	for index, unstructured := range unstructuredList.Items {
-		if ie.ShouldInclude(unstructured.GetName()) {
-			log.Debugf("Backup namespace %s due to namespace filters setting.", unstructured.GetName())
+func (r *itemCollector) backupNamespaces(
+	resource metav1.APIResource,
+	gv schema.GroupVersion,
+	gr schema.GroupResource,
+	preferredGVR schema.GroupVersionResource,
+	log logrus.FieldLogger,
+) ([]*kubernetesResource, error) {
+	resourceClient, err := r.dynamicFactory.ClientForGroupVersionResource(gv, resource, "")
+	if err != nil {
+		log.WithError(err).Error("Error getting dynamic client")
+		return nil, errors.WithStack(err)
+	}
+
+	unstructuredList, err := resourceClient.List(metav1.ListOptions{})
+	if err != nil {
+		log.WithError(errors.WithStack(err)).Error("error list namespaces")
+		return nil, errors.WithStack(err)
+	}
 
-			path, err := r.writeToFile(&unstructuredList.Items[index])
+	var singleSelector labels.Selector
+	var orSelectors []labels.Selector
+
+	if r.backupRequest.Backup.Spec.LabelSelector != nil {
+		var err error
+		singleSelector, err = metav1.LabelSelectorAsSelector(
+			r.backupRequest.Backup.Spec.LabelSelector)
+		if err != nil {
+			log.WithError(err).Errorf("Fail to convert backup LabelSelector %s into selector.",
+				metav1.FormatLabelSelector(r.backupRequest.Backup.Spec.LabelSelector))
+		}
+	}
+	if r.backupRequest.Backup.Spec.OrLabelSelectors != nil {
+		for _, ls := range r.backupRequest.Backup.Spec.OrLabelSelectors {
+			orSelector, err := metav1.LabelSelectorAsSelector(ls)
 			if err != nil {
-				log.WithError(err).Error("Error writing item to file")
-				continue
+				log.WithError(err).Errorf("Fail to convert backup OrLabelSelector %s into selector.",
+					metav1.FormatLabelSelector(ls))
 			}
+			orSelectors = append(orSelectors, orSelector)
+		}
+	}
 
-			items = append(items, &kubernetesResource{
-				groupResource: gr,
-				preferredGVR:  preferredGVR,
-				name:          unstructured.GetName(),
-				path:          path,
-			})
+	r.nsTracker.init(
+		unstructuredList.Items,
+		singleSelector,
+		orSelectors,
+		r.backupRequest.NamespaceIncludesExcludes,
+		log,
+	)
+
+	var items []*kubernetesResource
+
+	for index, unstructured := range unstructuredList.Items {
+		path, err := r.writeToFile(&unstructuredList.Items[index])
+		if err != nil {
+			log.WithError(err).Error("Error writing item to file")
```

Comment:
Would it be better to log the resource name?

Reply:
Added.